### DISCUSSION TITLE
prefetching,  additionalTypenames and debugging

### DISCUSSION THREAD

**Discussion by xjamundx** [urql-graphql/urql#2798] [Q&A] [answered]
I'm currently working on adding prefetching to a new urql application that relies on the document cache.

The pre-fetching happens using the syntax shown [on the website](https://formidable.com/open-source/urql/docs/basics/ui-patterns/#prefetching-data) like this:

```tsx
const context = { additionalTypenames: ["Book"] };
graphQLClient.query(ListSoldBooks, {}, { context }).toPromise()
```

This request is fired off before react is even `import()`d into the page and any of that happens. It kicks off right away. Later when we're referencing the query with this in react it uses the cache properly and doesn't repeat the call:

```tsx
// outside of the component
const context = { additionalTypenames: ["Book"] };

// in the component
const [{ fetching, data }] = const useListSoldBookQuery({ context })
```

All of that works beautifully, but the problem is when I go to fetch a create a new Book.

My mutation looks like this:

```tsx
// in an onSubmit callback
const additionalTypenames =  { additionalTypenames: ["Book"] };
const { error } = await createBook({ data }, additionalTypenames);
```

It does not seem to blow out the "Book" cache on mutation and when I go back to my book list the new book doesn't show up.

If I don't do the prefetching the cache invalidation works fine. Does manually querying not respect the additinalTypenames or did I just do something wrong?

---

Second, and related issue:

In the [cache exchange](https://github.com/FormidableLabs/urql/blob/7570219c37df5f84db153880ace701d9b3df9e72/packages/core/src/exchanges/cache.ts#L53) there are a number of `dispatchDebug` events, but none of them show up in my `subscribeToDebugTarget` event or in the dev tool. I was trying to use that to debug what was happening with caching, to no luck.

---

I'm using `urql@3.0.3`

**Comment by JoviDeCroock** [comment]
Quickly tried reproducing the [debug](https://codesandbox.io/s/gifted-leftpad-2enfg7?file=/src/index.js) issue you described but the subscription is working, what it could be however is that you are trying to log with `NODE_ENV=production` which would make it not log. If you are using pre-@urql/core v3.0.5 the source of the subscription is wrong.

For the `additionalTypenames` issue, are you using the same client to prefetch stuff? The behavior should be identical as the React layer does not do anything extra.

**Reply by xjamundx** [reply]
The logging issue is weird, because the fetch logs all come through, but nothing from any of the other exchanges. I removed my extra exchanges and went with the default and that didn't matter either. I can place breakpoints and know they're being called. At one point I `.toString()`'d the function that was being called and it seemed legit, not in production mode. I'll do a bit more digging on that.

**Reply by JoviDeCroock** [reply]
Have you checked your @urql/core version, that would make the source wrong as mentioned above

**Reply by xjamundx** [reply]
```
├── @urql/core@3.0.4
├─┬ @urql/devtools@2.0.3
│ └── @urql/core@3.0.4 deduped
├─┬ @urql/exchange-auth@1.0.0
│ └── @urql/core@3.0.4 deduped
└─┬ urql@3.0.3
  └── @urql/core@3.0.4 deduped
```

Checking!

**Reply by JoviDeCroock** [reply]
So upgrading to 3.0.5 should resolve that

**Reply by xjamundx** [reply]
Worked ✅

**Comment by xjamundx** [comment, marked as answer]
For the additionalTypename issue, here's some diving into the code to show the difference between what happens when I prefetch vs not. It's definitely the same client, because the caching works fine for reads, just doesn't bust on mutation.

![urql-debug](https://user-images.githubusercontent.com/246143/201330519-992240a6-6e57-48ca-97df-cf90669c6b6a.png)

**Reply by xjamundx** [reply]
BTW additionalTypenames are required because the way we do our queries we split our "Book" type into `NewBook` and `UsedBook` (not really, but conceptionally), and when a new book is added I want to make sure that both types are reset

**Reply by kitten** [reply]
Ok, so those lines are only for mutations, and not queries. However, for queries, the information for mutations is stored no matter whether a query stays active or not. If you don't see this being called, are you sure you're using the same client for both and your mutation is sent off correctly?

The latter you could verify in a quick'n'dirty way with the debugExchange.

**Reply by xjamundx** [reply]
Here's the broken case, will try without prefetch next to compare
![Screenshot 2022-11-11 at 11 37 48 am](https://user-images.githubusercontent.com/246143/201333413-3cc672ae-1329-4601-b50a-d6110f6abff0.png)

**Reply by xjamundx** [reply]
Literally the only difference is commenting out two lines of code to kick off the query early and this stuff start working again.
![Screenshot 2022-11-11 at 11 44 11 am](https://user-images.githubusercontent.com/246143/201334040-a6473293-ed31-4225-a1d8-e13dc8374f44.png)

**Reply by JoviDeCroock** [reply]
What are the typenames contained within said response and are the ones being invalidated by your third response in there?

**Reply by xjamundx** [reply]
And yeah when it works (no prefetch) the breakpoint fires on this [line 130 above](https://github.com/FormidableLabs/urql/blob/ef4b5b002d31e17d84fb41e5feb24d2200c9656f/packages/core/src/exchanges/cache.ts#L135):
`resultCache.delete(key)`

**Reply by xjamundx** [reply]
In both cases it says (prefetch and not):

> "The following typenames have been invalidated: EntityResponse, DataRequest"

The actual typenames in the response are in a list, and the actual typenames are `DataRequestRequesterDetails` essentially my `UsedBook` example

And in neither case does it say it's being invalidated.

**Reply by kitten** [reply]
Are you sure that both queries are the same? The document cache, since it's not normalised, requires you to have the same requests in both places. We don't have as much info as what you can see, but I'd stronglyu recommend adding the debugExchange so you can see results and operations: https://github.com/FormidableLabs/urql/blob/ef4b5b002d31e17d84fb41e5feb24d2200c9656f/packages/core/src/exchanges/debug.ts

**Reply by xjamundx** [reply]
I didn't understand what you meant by debug exchange (had the devtool one instead). Will add it and report back. Really appreciate your help here!!

**Reply by xjamundx** [reply]
Let me know if you can make any sense of this suggest anywhere I can debug. I can also try to get a proper repro case up on codesandbox. Thanks a ton!

working as expected:
![good-done](https://user-images.githubusercontent.com/246143/201339597-7fb173ef-11be-4729-b481-6bb0b6d49974.png)

not invalidating the cache after pre-fetching
![bad-done](https://user-images.githubusercontent.com/246143/201339613-a3372cd3-d438-49f8-91fb-f2fd767df33f.png)

**Reply by xjamundx** [reply]
I might be wrong, but I think the issue can be seen here in the cache exchange:

```tsx
      if (operation.kind === 'query' && response.data) {
          resultCache.set(operation.key, response);
          for (let i = 0; i < typenames.length; i++) {
            const typeName = typenames[i];
            let operations = operationCache.get(typeName);
            if (!operations)
              operationCache.set(typeName, (operations = new Set()));
            operations.add(operation.key);
          }
        }
```

For the pre-fetched queries we're not including the `additionalTypenames` here, but in the `useQuery`  case we are. Two screenshots:

**additionalTypenames added**
![Screenshot 2022-11-11 at 12 51 10 pm](https://user-images.githubusercontent.com/246143/201344248-daf65bda-a01f-4c1e-869c-ee7e02de5fe6.png)

**additionalTypenames not added**
![Screenshot 2022-11-11 at 12 50 45 pm](https://user-images.githubusercontent.com/246143/201344255-12fd7fd8-7e18-482d-92e6-80cb32c43599.png)

Let me show you what my queries look like exactly (with the books thing):

```tsx
    const context = { additionalTypenames: [AdditionalTypeNames.Book] };
    void graphQLClient.query(ListUsedBooksDocument, {}, { context }).toPromise();
```

And the code-generated version we're calling elsewhere:

```tsx
export function useListUseBooksQuery(options) {
  return Urql.useQuery({ query: Types.ListUsedBooksDocument, ...options });
};


// in the component
const [{ data, fetching }] = useListUseBooksQuery({ context });
```
Am I passing in context incorrectly in my manual query? That's kind of my guess. Or does it actually not pass them through?

**Reply by JoviDeCroock** [reply]
Typescript should be telling you that that query signature is incorrect.... `graphQLClient.query(ListUsedBooksDocument, {}, { ...context }).toPromise();`

**Reply by xjamundx** [reply]
Yeah ok it was me.

Needed:
```tsx
   const context = { additionalTypenames: [AdditionalTypeNames.Book] };
    void graphQLClient.query(ListUsedBooksDocument, {}, context).toPromise();
```
Instead of this:

```tsx
    void graphQLClient.query(ListUsedBooksDocument, {}, { context }).toPromise();
```

I feel like typescript let me down here, but I know for something vague like context it's hard to enforce what can and can't be in there. Ahh well.

Amazing help today. I learned a ton!

**Reply by xjamundx** [reply]
ha you solved it just before i did

**Reply by xjamundx** [reply]
Thank you both so much for your help 🙏🏻

**Comment by xjamundx** [comment]
Unfortunately, this didn't seem to work, but wondering if there's a similar TS way to make this more obvious:

```tsx
export interface OperationContext {
    [key: string]: any;
    context: never;
}
const x: OperationContenxt = {
    a: "B",
    context: "bbb",
};
```

Thanks again for pointing out the problem. It works now. We're porting from raw redux to urql and I didn't want to lose our pre-fetching goodness!
